mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRestClient.java

```diff
@@ -490,39 +503,296 @@ public TaskData getTaskData(TaskRepository repository, String taskId, IProgressM
+		    } else if (stateText.startsWith("reopened")) {
+			typeAttribute.setValue(ActivityType.REOPEN.toString());
+		    }
+
+		}
+	    }
+	    JsonArray labels = getIssueLabelEvents(matcher.group(1), matcher.group(2), OperationUtil.convert(monitor));
+	    if (labels != null) {
+		long lastLabelAt = 0;
+		TaskAttribute labelText = null;
+		ArrayList<String> added = new ArrayList<String>();
+		ArrayList<String> removed = new ArrayList<String>();
+		for (JsonElement stateElem : labels) {
+		    JsonObject label = (JsonObject) stateElem;
+		    Date date = GitlabTaskAttributeMapper.parseDate(label.get("created_at").getAsString());
+
+		    if (lastLabelAt != date.getTime()) {
+			lastLabelAt = date.getTime();
+
+			buildLableText(labelText, added, removed);
+
+			added.clear();
+			removed.clear();
+
+			TaskAttribute taskAttribute = result.getRoot()
+				.createAttribute(GitlabCoreActivator.PREFIX_ACTIVITY + date.getTime());
+			taskAttribute.getMetaData().setType(GitlabCoreActivator.ATTRIBUTE_TYPE_ACTIVITY);
+			TaskAttribute child = DefaultTaskSchema.getField(TaskAttribute.COMMENT_DATE)
+				.createAttribute(taskAttribute);
+			child.setValue(DateFormat.getDateTimeInstance().format(date));
+
+			child = DefaultTaskSchema.getField(TaskAttribute.COMMENT_AUTHOR).createAttribute(taskAttribute);
+			child.setValue(label.get("user").getAsJsonObject().get("name").getAsString());
+
+			TaskAttribute typeAttribute = taskAttribute
+				.createAttribute(GitlabCoreActivator.ATTRIBUTE_TYPE_ACTIVITY);
+			typeAttribute.setValue(ActivityType.LABEL.toString());
+			labelText = DefaultTaskSchema.getField(TaskAttribute.COMMENT_TEXT)
+				.createAttribute(taskAttribute);
+			taskAttribute.getMetaData().putValue(TaskAttribute.META_ASSOCIATED_ATTRIBUTE_ID,
+				labelText.getId());
+		    }
+		    if (label.getAsJsonObject().get("action").getAsString().equals("add")) {
+			added.add(label.get("label").getAsJsonObject().get("name").getAsString());
+		    }
+		    if (label.getAsJsonObject().get("action").getAsString().equals("remove")) {
+			removed.add(label.get("label").getAsJsonObject().get("name").getAsString());
+		    }
+
+		}
+		buildLableText(labelText, added, removed);
+
+	    }
+
 	    config.updateProductOptions(result);
 	}
+	if (TRACE) {
+	    endTime = Calendar.getInstance().getTimeInMillis();
+	    thread = Thread.currentThread();
+	    System.out.println(thread.getName() + " getTaskData taken: " + (endTime - startTime) + " ms "
+		    + repository.getRepositoryUrl() + " id " + taskId);
+	}
 	return result;
     }
 
+    private void buildLableText(TaskAttribute labelText, ArrayList<String> added, ArrayList<String> removed) {
+	if (labelText != null) {
+	    String text = "";
+	    if (added.size() > 0) {
+		text += "added ";
+		text += String.join(", ", added);
+		if (added.size() > 1) {
+		    text += " labels";
+		} else {
+		    text += " label";
+		}
+	    }
+	    if (removed.size() > 0) {
+		if (text.length() == 0) {
```

Comment:
`text += text.length() == 0 ? "removed" : "and removed"`

Reply:
Done with new commit.